packages/next/src/server/app-render/dynamic-rendering.ts

```diff
@@ -117,6 +117,47 @@ export function markCurrentScopeAsDynamic(
   }
 }
 
+/**
+ * This function communicates that some dynamic path parameter was read. This
+ * differs from the more general `trackDynamicDataAccessed` in that it is will
+ * not error when `dynamic = "error"` is set.
+ *
+ * @param store The static generation store
+ * @param expression The expression that was accessed dynamically
+ */
+export function trackFallbackParamAccessed(
+  store: StaticGenerationStore,
+  expression: string
+): void {
+  if (store.isUnstableCacheCallback) {
+    throw new Error(
+      `Route ${store.route} used "${expression}" inside a function cached with "unstable_cache(...)". Accessing Dynamic data sources inside a cache scope is not supported. If you need this data inside a cached function use "${expression}" outside of the cached function and pass the required dynamic data in as an argument. See more info here: https://nextjs.org/docs/app/api-reference/functions/unstable_cache`
+    )
```

Comment:
For cookies() and headers() we error if it's inside a cache scope even if we're doing a fully dynamic render. Is it even possible to define a cache function during a render? If you can't then it's not really possible to ready params from inside the function because we should be serializing the args and keying that and this would trigger dynamic from outside.

If you can construct a cache function from inside a render then we actually need this error to throw all the time, not just on fallback renders because otherwise you'll never know you misconfigured your app in dev mode and will only see it error when you build and even then only if you have ppr fallbacks turned on.

This is a bit different than cookies() and headers() because with those they're just global functions so it's trivial to get them inside the cache function even if it is defined outside of render.

But back to my point. We probably want this tracking 100% of the time and we just need to ensure it doesn't postpone if there are unknown params.